// ngx-highlightjs/bundles/ngx-highlightjs.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs/BehaviorSubject'), require('rxjs/observable/from'), require('rxjs/add/operator/map'), require('rxjs/add/operator/take'), require('rxjs/add/operator/filter')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', 'rxjs/BehaviorSubject', 'rxjs/observable/from', 'rxjs/add/operator/map', 'rxjs/add/operator/take', 'rxjs/add/operator/filter'], factory) :
	(factory((global.ng = global.ng || {}, global.ng.ngxHighlightjs = global.ng.ngxHighlightjs || {}),global.ng.core,global.Rx,global.Rx));
}(this, (function (exports,_angular_core,rxjs_BehaviorSubject,rxjs_observable_from) { 'use strict';

var OPTIONS = new _angular_core.InjectionToken('OPTIONS');
var HighlightService = (function () {
    /**
     * @param {?} options
     */
    function HighlightService(options) {
        this.options = {
            theme: 'github',
            path: 'assets/lib/hljs',
            auto: true
        };
        /**
         * hljs script on load event
         */
        this.ready$ = new rxjs_BehaviorSubject.BehaviorSubject(false);
        this.options = Object.assign({}, this.options, options);
        /** Load hljs script and style only once */
        this.loadScript();
        this.loadTheme();
    }
    /**
     * @return {?}
     */
    HighlightService.prototype.loadScript = function () {
        var _this = this;
        var /** @type {?} */ script = document.createElement('script');
        script.async = true;
        script.type = 'text/javascript';
        script.onload = function () {
            _this.ready$.next(true);
        };
        script.src = this.options.path + "/highlight.pack.js";
        document.head.appendChild(script);
    };
    /**
     * @return {?}
     */
    HighlightService.prototype.loadTheme = function () {
        var /** @type {?} */ style = document.createElement('link');
        style.rel = 'stylesheet';
        style.type = 'text/css';
        style.href = this.options.path + "/styles/" + this.options.theme + ".css";
        document.head.appendChild(style);
    };
    return HighlightService;
}());
HighlightService.decorators = [
    { type: _angular_core.Injectable },
];
/**
 * @nocollapse
 */
HighlightService.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: _angular_core.Optional }, { type: _angular_core.Inject, args: [OPTIONS,] },] },
]; };
/**
 * There are 2 ways to highlight code
 1 - Directly using the code input (default)
 2 - Indirectly using element text content (used of code input is undefined)
 */
var HighlightDirective = (function () {
    /**
     * @param {?} el
     * @param {?} renderer
     * @param {?} hl
     */
    function HighlightDirective(el, renderer, hl) {
        this.renderer = renderer;
        this.hl = hl;
        this.hljsClass = true;
        this.el = el.nativeElement;
    }
    Object.defineProperty(HighlightDirective.prototype, "setCode", {
        /**
         * Highlight code directly
         * @param {?} code
         * @return {?}
         */
        set: function (code) {
            var _this = this;
            this.code = code;
            this.hl.ready$.filter(function (ready) { return ready; }).take(1)
                .subscribe(function () { return _this.highlightElement(_this.el, code); });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    HighlightDirective.prototype.ngOnInit = function () {
        var _this = this;
        /** If code is undefined, highlight using element text content */
        if (!this.code && this.hl.options.auto) {
            this.hl.ready$
                .filter(function (ready) { return ready; }).take(1)
                .subscribe(function () {
                _this.highlightTextContent();
                /** Highlight when text content changes */
                _this.domObs = new MutationObserver(function () { return _this.highlightTextContent(); });
                _this.domObs.observe(_this.el, { childList: true, subtree: true });
            });
        }
    };
    /**
     * Highlight using element text content
     * @return {?}
     */
    HighlightDirective.prototype.highlightTextContent = function () {
        if (!this.highlight) {
            /** <code highlight [textContent]="code"></code> */
            if (this.el.tagName.toLowerCase() === 'code') {
                this.highlightElement(this.el, this.el.innerText.trim());
            }
            else {
                console.warn("[HighlightDirective]: Use 'highlight' on <code> elements only");
            }
        }
        else if (this.highlight === 'all') {
            /** <div highlight="all">
             *    <pre><code [textContent]="code"></code></pre>
             *    <pre><code [textContent]="code"></code></pre>
             *  </div>
             */
            this.highlightChildren(this.el, 'pre code');
        }
        else {
            /** <div highlight="section code">
             *    <section><code [textContent]="code"></code></section>
             *    <section><code [textContent]="code"></code></section>
             *  </div>
             */
            this.highlightChildren(this.el, this.highlight);
        }
    };
    /**
     * Highlight single element
     * @param {?} el
     * @param {?} code
     * @return {?}
     */
    HighlightDirective.prototype.highlightElement = function (el, code) {
        var /** @type {?} */ highlightedCode = hljs.highlightAuto(code).value;
        if (highlightedCode !== el.innerHTML) {
            this.renderer.setProperty(el, 'innerHTML', highlightedCode);
        }
    };
    /**
     * Highlight children
     * @param {?} el
     * @param {?} selector
     * @return {?}
     */
    HighlightDirective.prototype.highlightChildren = function (el, selector) {
        var _this = this;
        var /** @type {?} */ codeElements = el.querySelectorAll(selector);
        /** highlight all children with the same selector */
        rxjs_observable_from.from(codeElements)
            .filter(function (code) { return code.childNodes.length === 1 && code.childNodes[0].nodeName === '#text'; })
            .map(function (codeElement) {
            _this.highlightElement(codeElement, codeElement.innerText.trim());
        })
            .take(1)
            .subscribe();
    };
    /**
     * @return {?}
     */
    HighlightDirective.prototype.ngOnDestroy = function () {
        /** Disconnect MutationObserver */
        if (!this.code && this.hl.options.auto) {
            this.domObs.disconnect();
        }
    };
    return HighlightDirective;
}());
HighlightDirective.decorators = [
    { type: _angular_core.Directive, args: [{
                selector: '[highlight]'
            },] },
];
/**
 * @nocollapse
 */
HighlightDirective.ctorParameters = function () { return [
    { type: _angular_core.ElementRef, },
    { type: _angular_core.Renderer2, },
    { type: HighlightService, },
]; };
HighlightDirective.propDecorators = {
    'highlight': [{ type: _angular_core.Input },],
    'setCode': [{ type: _angular_core.Input, args: ['code',] },],
    'hljsClass': [{ type: _angular_core.HostBinding, args: ['class.hljs',] },],
};
/**
 * @param {?} options
 * @return {?}
 */
function HighlightFactory(options) {
    return new HighlightService(options);
}
var HighlightModule = (function () {
    function HighlightModule() {
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    HighlightModule.forRoot = function (options) {
        return {
            ngModule: HighlightModule,
            providers: [
                { provide: OPTIONS, useValue: options },
                {
                    provide: HighlightService,
                    useFactory: HighlightFactory,
                    deps: [OPTIONS]
                }
            ]
        };
    };
    return HighlightModule;
}());
HighlightModule.decorators = [
    { type: _angular_core.NgModule, args: [{
                declarations: [HighlightDirective],
                exports: [HighlightDirective]
            },] },
];
/**
 * @nocollapse
 */
HighlightModule.ctorParameters = function () { return []; };

exports.HighlightModule = HighlightModule;
exports.ɵb = HighlightDirective;
exports.ɵa = HighlightFactory;
exports.ɵd = HighlightService;
exports.ɵc = OPTIONS;

Object.defineProperty(exports, '__esModule', { value: true });

})));
